source/common/compressor/compressor_impl.cc

```diff
@@ -0,0 +1,91 @@
+#include "common/compressor/compressor_impl.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Compressor {
+
+CompressorImpl::CompressorImpl()
+    : zstream_ptr_(new z_stream(), [](z_stream* z) {
+        deflateEnd(z);
+        delete z;
+      }) {
+  zstream_ptr_->zalloc = Z_NULL;
+  zstream_ptr_->zfree = Z_NULL;
+  zstream_ptr_->opaque = Z_NULL;
+}
+
+CompressorImpl::~CompressorImpl() {}
+
+void CompressorImpl::init(CompressionLevel comp_level, CompressionStrategy comp_strategy,
+                          int window_bits, uint8_t memory_level = 8) {
+  if (initialized_) {
+    throw EnvoyException("Compressor can only be initialized once");
+  }
+
+  const int res = deflateInit2(zstream_ptr_.get(), comp_level, Z_DEFLATED, window_bits,
+                               memory_level, comp_strategy);
+
+  if (res != Z_OK) {
+    // TODO: error mapping
+    throw EnvoyException(fmt::format("Compressor initialization error: '{}'", res));
+  }
+
+  initialized_ = true;
+}
+
+void CompressorImpl::setChunk(uint64_t chunk) { chunk_ = chunk; }
+
+void CompressorImpl::finish(Buffer::Instance& output_buffer) {
+  process(output_buffer, Z_SYNC_FLUSH);
+  commit(output_buffer);
+}
+
+void CompressorImpl::compress(const Buffer::Instance& intput_buffer,
+                              Buffer::Instance& output_buffer) {
+  if (zstream_ptr_->total_out == 0) {
+    reserve(output_buffer);
+  }
+
+  const uint64_t num_slices = intput_buffer.getRawSlices(nullptr, 0);
+  Buffer::RawSlice slices[num_slices];
+  intput_buffer.getRawSlices(slices, num_slices);
+
+  for (const Buffer::RawSlice& input_slice : slices) {
+    zstream_ptr_->avail_in = input_slice.len_;
+    zstream_ptr_->next_in = static_cast<Bytef*>(input_slice.mem_);
+    process(output_buffer, Z_NO_FLUSH);
+  }
+}
+
+void CompressorImpl::process(Buffer::Instance& output_buffer, int flush_state) {
+  do {
+    if (zstream_ptr_->avail_out == 0) {
+      commit(output_buffer);
+      reserve(output_buffer);
+    }
+    const int res = deflate(zstream_ptr_.get(), flush_state);
+    if (res != Z_OK) {
+      // TODO: error mapping
```

Comment:
TODO is fine, just put your GH ID in there.

Reply:
👍